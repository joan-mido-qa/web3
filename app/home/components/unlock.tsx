import { ClipboardEvent, FormEvent, useState } from "react";
import * as bip39 from "@scure/bip39";
import { HDKey } from "@scure/bip32";
import Web3 from "web3";
import Wallet from "web3-eth-accounts";

interface Props {
  web3: Web3;
  isLogged: boolean;
  onUnlock: (value: Wallet.Wallet) => void;
}

export default function Unlock({ web3, isLogged, onUnlock }: Props) {
  const [mnemonic, setMnemonic] = useState(emptyMnemonic());
  const [passphrase, setPassphrase] = useState("");

  function emptyMnemonic(): Map<number, string> {
    return new Map(Array.from({ length: 12 }).map((_, i) => [i, ""]));
  }

  function onPaste(event: ClipboardEvent<HTMLInputElement>) {
    event.preventDefault();

    const newMnemonic = emptyMnemonic();

    const pastedMnemonic = event.clipboardData.getData("Text").split(" ");

    newMnemonic.forEach((_, key) => {
      if (newMnemonic.has(key)) {
        newMnemonic.set(key, pastedMnemonic[key]);
      }
    });

    setMnemonic(newMnemonic);
  }

  function getAccount(key: HDKey, index: number) {
    return web3.eth.accounts.privateKeyToAccount(
      key.deriveChild(index).privateKey!
    );
  }

  async function onSubmit(event: FormEvent<HTMLFormElement>) {
    event.preventDefault();

    if (!isLogged) {
      event.preventDefault();

      const seed = await bip39.mnemonicToSeed(
        Array.from(mnemonic.values())
          .reduce((mnemonic, word) => `${mnemonic} ${word}`, "")
          .trim(),
        passphrase
      );

      const hdKey = HDKey.fromMasterSeed(seed).derive("m/44'/60'/0'/0");

      for (let index = 0; index < 10; index++) {
        web3.eth.accounts.wallet.add(getAccount(hdKey, index));
      }

      await web3.eth.accounts.wallet.save(passphrase);

      setMnemonic(emptyMnemonic());
    }

    onUnlock(await web3.eth.accounts.wallet.load(passphrase));

    setPassphrase("");
  }

  return !isLogged ? (
    <>
      <div className='text-white flex flex-col items-center justify-center'>
        <h1 className='text-4xl font-bold'>Wallet Access</h1>
        <p className='mt-4 text-s'>Mnemonic & Password</p>
      </div>
      <form
        onSubmit={onSubmit}
        className='flex flex-col justify-around items-center w-[750px] h-[375px] p-2'
      >
        <div className='flex flex-row justify-center items-center flex-wrap w-full'>
          {Array.from(mnemonic.entries()).map(([key, value]) => (
            <input
              placeholder={`Word ${key + 1}`}
              type='text'
              className='m-[5px] py-2 px-4 bg-gray-800 text-white rounded-md focus:outline-none baisis-[calc(100%/3) - 5]'
              key={`word-${key}`}
              value={value}
              onChange={(e) =>
                setMnemonic(new Map(mnemonic.set(key, e.target.value)))
              }
              onPaste={(e) => onPaste(e)}
            />
          ))}
        </div>
        <input
          placeholder='Passphrase'
          type='text'
          className='py-2 px-4 bg-gray-800 text-white rounded-md focus:outline-none'
          onChange={(e) => setPassphrase(e.target.value)}
          value={passphrase}
        />
        <button
          type='submit'
          className='bg-blue-500 py-1.5 px-8 text-white rounded-md hover:bg-blue-600 focus:outline-none'
        >
          Login
        </button>
      </form>
    </>
  ) : (
    <>
      <div className='justify-center items-center flex fixed inset-0 z-50 outline-none focus:outline-none'>
        <form
          onSubmit={onSubmit}
          className='p-2 flex flex-row items-around bg-white rounded-md outline-none focus:outline-none'
        >
          <input
            placeholder='Passphrase'
            type='text'
            className='p-3 m-1 bg-gray-800 text-white rounded-md focus:outline-none'
            onChange={(e) => setPassphrase(e.target.value)}
            value={passphrase}
          />
          <button
            className='bg-blue-500 text-white hover:bg-blue-600 font-bold uppercase text-sm p-3 m-1 rounded shadow hover:shadow-lg outline-none focus:outline-none mr-1 mb-1 ease-linear transition-all duration-150'
            type='submit'
          >
            Unlock
          </button>
        </form>
      </div>
    </>
  );
}
