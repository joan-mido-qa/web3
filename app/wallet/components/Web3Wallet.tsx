"use client";

import { FormEvent, useEffect, useState } from "react";
import Web3 from "web3";
import Wallet from "web3-eth-accounts";
import { GrTransaction } from "react-icons/gr";
import { GrDocumentUpload } from "react-icons/gr";

interface Props {
  web3: Web3;
  accounts: Wallet.Wallet;
}

export default function Web3Wallet({ web3, accounts }: Props) {
  const [isSendDisplayed, setDisplaySend] = useState<boolean>(false);

  const [balance, setBalance] = useState<string>("0");
  const [account, setAccount] = useState<Wallet.Web3Account>(accounts.get(0)!);

  const updateBalance = () =>
    web3.eth
      .getBalance(account.address)
      .then((balance) => setBalance(Web3.utils.fromWei(balance, "ether")));

  useEffect(() => {
    const sub = web3.eth.subscribe("newBlockHeaders").then((sub) => {
      sub.on("data", async () => await updateBalance());
      return sub;
    });

    return () => {
      sub.then((sub) => sub.unsubscribe());
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    updateBalance();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [account]);

  const fromAddress = (address: string) =>
    accounts.find((acc) => acc.address == address)!;

  return (
    <>
      <div className='flex flex-col text-white justify-around items-center h-[250px]'>
        <span className='text-5xl'>{balance} ETH</span>
        <select
          onChange={(e) => setAccount(fromAddress(e.target.value))}
          className='m-[5px] py-2 px-4 bg-gray-800 text-white rounded-md focus:outline-none'
        >
          {accounts.map((acc, i) => (
            <option
              key={`account-${i}`}
              className='bg-gray-800'
              value={acc.address}
            >
              {`${acc.address.substring(0, 5)}...${acc.address.substring(acc.address.length - 5)}`}
            </option>
          ))}
        </select>
        <div className='flex'>
          <div className='flex flex-col items-center mx-2'>
            <button
              onClick={() => setDisplaySend(true)}
              className='bg-blue-500 hover:bg-blue-600 text-white text-l p-2 m-2 rounded-full'
            >
              <GrTransaction />
            </button>
            <span>Send</span>
          </div>
          <div className='flex flex-col items-center mx-2'>
            <button className='bg-blue-500 hover:bg-blue-600 text-white text-l p-2 m-2 rounded-full'>
              <GrDocumentUpload />
            </button>
            <span>Upload</span>
          </div>
        </div>
      </div>
      {isSendDisplayed ? (
        <SendModal
          web3={web3}
          fromAccount={account}
          onSend={() => setDisplaySend(false)}
        />
      ) : null}
    </>
  );
}

interface SendProps {
  web3: Web3;
  fromAccount: Wallet.Web3Account;
  onSend: () => void;
}

function SendModal({ web3, fromAccount, onSend }: SendProps) {
  const [toAdress, setAdress] = useState<string>("");
  const [amount, setAmount] = useState<string>("");

  async function onSubmit(event: FormEvent<HTMLFormElement>) {
    event.preventDefault();

    const nonce = await web3.eth.getTransactionCount(
      fromAccount.address,
      "latest"
    );
    const gasPrice = await web3.eth.getGasPrice();

    const transaction = {
      to: toAdress,
      value: web3.utils.toWei(amount!, "ether"),
      gasPrice: gasPrice,
      nonce: nonce,
    };

    const gas = await web3.eth.estimateGas(transaction);

    const signedTransaction = await fromAccount.signTransaction({
      ...transaction,
      gas,
    });

    await web3.eth.sendSignedTransaction(signedTransaction.rawTransaction);

    onSend();
  }

  return (
    <>
      <div className='justify-center items-center flex fixed inset-0 z-50 outline-none focus:outline-none'>
        <form
          onSubmit={onSubmit}
          className='p-2 flex flex-col items-around bg-white rounded-md outline-none focus:outline-none'
        >
          <input
            placeholder='Address'
            type='text'
            className='p-3 m-1 bg-gray-800 text-white rounded-md focus:outline-none'
            onChange={(e) => setAdress(e.target.value)}
            value={toAdress}
          />
          <input
            placeholder='Amount (Eth)'
            type='text'
            className='p-3 m-1 bg-gray-800 text-white rounded-md focus:outline-none'
            onChange={(e) => setAmount(e.target.value)}
            value={amount}
          />
          <button
            className='bg-blue-500 text-white hover:bg-blue-600 font-bold uppercase text-sm p-3 m-1 rounded shadow hover:shadow-lg outline-none focus:outline-none mr-1 mb-1 ease-linear transition-all duration-150'
            type='submit'
          >
            Send
          </button>
        </form>
      </div>
      <div className='opacity-50 fixed inset-0 z-40 bg-black'></div>
    </>
  );
}
