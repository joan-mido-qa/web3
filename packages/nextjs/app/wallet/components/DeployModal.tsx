"use client";

import { Web3Context } from "@/app/Web3Provider";
import { FormEvent, useState } from "react";
import Wallet from "web3-eth-accounts";

interface DeployProps {
	web3: Web3Context;
	fromAccount: Wallet.Web3Account;
	onError: (err: Error) => void;
	onAccept: () => void;
}

const NoContractFileSelectedError = new Error("Select a file containing the Contract ABI and Bytecode");

export default function DeployModal({ web3, fromAccount, onError, onAccept }: DeployProps) {
	const [files, setFiles] = useState<FileList | null>();
	const [address, setAddress] = useState<string>("");

	async function onSubmit(event: FormEvent<HTMLFormElement>) {
		event.preventDefault();

		if (!files) {
			return onError(NoContractFileSelectedError);
		}

		const reader = new FileReader();

		reader.onload = async (event) => {
			let contract;

			try {
				contract = JSON.parse(event.target!.result as string);
			} catch (e) {
				return onError(NoContractFileSelectedError);
			}

			const isContract = "abi" in contract && "bytecode" in contract;

			if (!isContract) return onError(NoContractFileSelectedError);

			await web3
				.deploySignedContratc(fromAccount, contract.abi, contract.bytecode, [["Option 1", "Option 2", "Option 3Q"]])
				.then((recipt) => setAddress(recipt.contractAddress!))
				.catch((e) => onError(e));
		};

		reader.readAsText(files[0]);
	}

	if (address) {
		return (
			<>
				<div className="justify-center items-center flex fixed inset-0 z-50 outline-none focus:outline-none">
					<div className="p-2 flex flex-col items-around bg-white rounded-md outline-none focus:outline-none">
						<span data-testid="contract-address" className="text-blue-500 font-bold text-sm p-3 m-1">
							{address}
						</span>
						<button
							className="bg-blue-500 text-white hover:bg-blue-600 font-bold uppercase text-sm p-3 m-1 rounded shadow hover:shadow-lg outline-none focus:outline-none ease-linear transition-all duration-150"
							onClick={onAccept}
						>
							Ok
						</button>
					</div>
				</div>
				<div className="opacity-50 fixed inset-0 z-40 bg-black"></div>
			</>
		);
	}

	return (
		<>
			<div className="justify-center items-center flex fixed inset-0 z-50 outline-none focus:outline-none">
				<form
					onSubmit={onSubmit}
					className="p-2 flex flex-col items-around bg-white rounded-md outline-none focus:outline-none"
				>
					<label className="block">
						<input
							type="file"
							className="m-1 block text-sm text-gray-500 file:me-4 file:py-2 file:px-4 file:rounded-lg file:border-0 file:text-sm file:font-semibold file:bg-blue-600 file:text-white  hover:file:bg-blue-700 file:disabled:opacity-50 file:disabled:pointer-events-none  dark:file:bg-blue-500 dark:hover:file:bg-blue-400"
							onChange={(event) => setFiles(event.target.files)}
							accept="application/json"
							data-testid="file-input"
						/>
					</label>
					<button
						className="bg-blue-500 text-white hover:bg-blue-600 font-bold uppercase text-sm p-3 m-1 rounded shadow hover:shadow-lg outline-none focus:outline-none ease-linear transition-all duration-150"
						type="submit"
					>
						Deploy
					</button>
				</form>
			</div>
			<div className="opacity-50 fixed inset-0 z-40 bg-black"></div>
		</>
	);
}
