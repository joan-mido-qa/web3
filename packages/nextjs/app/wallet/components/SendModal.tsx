"use client";

import { FormEvent, useState } from "react";
import Wallet from "web3-eth-accounts";
import { Web3Context } from "@/app/Web3Provider";

interface SendProps {
  web3: Web3Context;
  fromAccount: Wallet.Web3Account;
  onError: (err: Error) => void;
  onSend: () => void;
}

export default function SendModal({ web3, fromAccount, onSend, onError }: SendProps) {
  const [toAdress, setAdress] = useState<string>("");
  const [amount, setAmount] = useState<string>("");

  async function onSubmit(event: FormEvent<HTMLFormElement>) {
    event.preventDefault();

    await web3
      .sendSignedTransaction(fromAccount, toAdress, amount)
      .then(() => onSend())
      .catch((e) => onError(e));
  }

  return (
    <>
      <div className='justify-center items-center flex fixed inset-0 z-50 outline-none focus:outline-none'>
        <form
          onSubmit={onSubmit}
          className='p-2 flex flex-col items-around bg-white rounded-md outline-none focus:outline-none'>
          <input
            placeholder='Address'
            type='text'
            className='p-3 m-1 bg-gray-800 text-white rounded-md focus:outline-none'
            onChange={(e) => setAdress(e.target.value)}
            value={toAdress}
          />
          <input
            placeholder='Amount (Eth)'
            type='text'
            className='p-3 m-1 bg-gray-800 text-white rounded-md focus:outline-none'
            onChange={(e) => setAmount(e.target.value)}
            value={amount}
          />
          <button
            className='bg-blue-500 text-white hover:bg-blue-600 font-bold uppercase text-sm p-3 m-1 rounded shadow hover:shadow-lg outline-none focus:outline-none ease-linear transition-all duration-150'
            type='submit'>
            Send
          </button>
        </form>
      </div>
      <div className='opacity-50 fixed inset-0 z-40 bg-black'></div>
    </>
  );
}
