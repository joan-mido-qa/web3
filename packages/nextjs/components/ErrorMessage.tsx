interface Props {
  message: string;
}

export default function ErrorMessage({ message }: Props) {
  return (
    <div className="z-50 absolute bottom-5 left-5 bg-red-100 border-l-4 border-red-500 text-red-700 p-4" role="alert">
      {message}
    </div>
  );
}
