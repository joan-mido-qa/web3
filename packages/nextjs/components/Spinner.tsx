export default function Spinner() {
	return (
		<div className="absolute right-1/2 bottom-1/2  transform translate-x-1/2 translate-y-1/2 ">
			<div className="border-t-transparent border-solid animate-spin  rounded-full border-blue-500 border-8 h-40 w-40"></div>
		</div>
	);
}
