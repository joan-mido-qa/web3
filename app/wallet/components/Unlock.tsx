import { ClipboardEvent, FormEvent, useState } from "react";
import * as bip39 from "@scure/bip39";
import { HDKey } from "@scure/bip32";
import Web3 from "web3";
import Wallet from "web3-eth-accounts";
import ErrorMessage from "./ErrorMessage";

interface Props {
  web3: Web3;
  isLogged: boolean;
  onUnlock: (value: Wallet.Wallet) => void;
}

export default function Unlock({ web3, isLogged, onUnlock }: Props) {
  const [mnemonic, setMnemonic] = useState(emptyMnemonic());
  const [passphrase, setPassphrase] = useState("");
  const [error, setError] = useState("");

  function emptyMnemonic(): Map<number, string> {
    return new Map(Array.from({ length: 12 }).map((_, i) => [i, ""]));
  }

  function onPaste(event: ClipboardEvent<HTMLInputElement>) {
    event.preventDefault();

    const newMnemonic = emptyMnemonic();

    const pastedMnemonic = event.clipboardData.getData("Text").split(" ");

    newMnemonic.forEach((_, key) => {
      if (newMnemonic.has(key)) {
        newMnemonic.set(key, pastedMnemonic[key]);
      }
    });

    setMnemonic(newMnemonic);
  }

  function getAccount(key: HDKey, index: number) {
    return web3.eth.accounts.privateKeyToAccount(key.deriveChild(index).privateKey!);
  }

  const getErrorMessage = (error: unknown) => (error instanceof Error ? error.message : String(error));

  async function onSubmit(event: FormEvent<HTMLFormElement>) {
    event.preventDefault();

    try {
      if (!isLogged) {
        const seed = bip39.mnemonicToSeedSync(
          Array.from(mnemonic.values())
            .reduce((mnemonic, word) => `${mnemonic} ${word}`, "")
            .trim()
        );
        const hdKey = HDKey.fromMasterSeed(seed).derive("m/44'/60'/0'/0");

        for (let index = 0; index < 10; index++) {
          web3.eth.accounts.wallet.add(getAccount(hdKey, index));
        }

        setMnemonic(emptyMnemonic());

        await web3.eth.accounts.wallet.save(passphrase);
      } else {
        await web3.eth.accounts.wallet.load(passphrase);
      }
    } catch (e) {
      return setError(getErrorMessage(e));
    } finally {
      setPassphrase("");
    }

    onUnlock(web3.eth.accounts.wallet);
  }

  return (
    <>
      {!isLogged ? (
        <>
          <div className='text-white flex flex-col items-center justify-center'>
            <h1 className='text-4xl font-bold'>Wallet Access</h1>
            <p className='mt-4 text-s'>Mnemonic & Password</p>
          </div>
          <form onSubmit={onSubmit} className='flex flex-col justify-around items-center w-[750px] h-[375px] p-2'>
            <div className='flex flex-row justify-center items-center flex-wrap w-full'>
              {Array.from(mnemonic.entries()).map(([key, value]) => (
                <input
                  placeholder={`Word ${key + 1}`}
                  type='text'
                  className='m-[5px] py-2 px-4 bg-gray-800 text-white rounded-md focus:outline-none baisis-[calc(100%/3) - 5]'
                  key={`word-${key}`}
                  value={value}
                  onChange={(e) => setMnemonic(new Map(mnemonic.set(key, e.target.value)))}
                  onPaste={(e) => onPaste(e)}
                />
              ))}
            </div>
            <input
              placeholder='Passphrase'
              type='text'
              className='py-2 px-4 bg-gray-800 text-white rounded-md focus:outline-none'
              onChange={(e) => setPassphrase(e.target.value)}
              value={passphrase}
            />
            <button
              type='submit'
              className='bg-blue-500 py-1.5 px-8 text-white rounded-md hover:bg-blue-600 focus:outline-none'>
              Login
            </button>
          </form>
        </>
      ) : (
        <>
          <div className='justify-center items-center flex fixed inset-0 z-50 outline-none focus:outline-none'>
            <form
              onSubmit={onSubmit}
              className='p-2 flex flex-row items-around bg-white rounded-md outline-none focus:outline-none'>
              <input
                placeholder='Passphrase'
                type='text'
                className='p-3 m-1 bg-gray-800 text-white rounded-md focus:outline-none'
                onChange={(e) => setPassphrase(e.target.value)}
                value={passphrase}
              />
              <button
                className='bg-blue-500 text-white hover:bg-blue-600 font-bold uppercase text-sm p-3 m-1 rounded shadow hover:shadow-lg outline-none focus:outline-none mr-1 mb-1 ease-linear transition-all duration-150'
                type='submit'>
                Unlock
              </button>
            </form>
          </div>
        </>
      )}
      {error ? <ErrorMessage message={error} /> : null}
    </>
  );
}
