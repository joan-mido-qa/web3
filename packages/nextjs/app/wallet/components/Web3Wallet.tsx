"use client";

import { FormEvent, useCallback, useEffect, useRef, useState } from "react";
import Web3 from "web3";
import Wallet from "web3-eth-accounts";
import { GrTransaction } from "react-icons/gr";
import { GrDocumentUpload } from "react-icons/gr";
import ErrorMessage from "./ErrorMessage";

interface Props {
  web3: Web3;
  wallet: Wallet.Wallet;
}

export default function Web3Wallet({ web3, wallet }: Props) {
  const [isSendDisplayed, setDisplaySend] = useState<boolean>(false);

  const [balance, setBalance] = useState<string>("0");
  const [account, setAccount] = useState<Wallet.Web3Account>(wallet.get(0)!);
  const [error, setError] = useState("");

  const prevBalanceRef = useRef("0");

  const updateBalance = useCallback(
    () =>
      web3.eth.getBalance(account.address, "latest").then((balance) => {
        const ethBalance = Web3.utils.fromWei(balance, "ether");
        if (ethBalance !== prevBalanceRef.current) {
          prevBalanceRef.current = ethBalance;
          setBalance(ethBalance);
        }
      }),
    [account]
  );

  useEffect(() => {
    const sub = web3.eth.subscribe("newBlockHeaders").then((sub) => {
      sub.on("data", () => updateBalance());
      // Avoid Unhandled Errors
      sub.on("error", () => {});
      return sub;
    });

    return () => {
      sub.then((sub) => sub.unsubscribe());
    };
  }, [updateBalance]);

  useEffect(() => {
    updateBalance();
  }, [updateBalance]);

  function setErrorMessage(err?: string) {
    if (err) {
      setError(err);
      setTimeout(() => {
        setError("");
      }, 3000);
    }
  }

  const fromAddress = (address: string) => wallet.find((acc) => acc.address == address)!;

  return (
    <>
      <div className='flex flex-col text-white justify-around items-center h-[250px]'>
        <span data-testid='eth-balance' className='text-5xl'>
          {balance} ETH
        </span>
        <select
          onChange={(e) => setAccount(fromAddress(e.target.value))}
          className='m-[5px] py-2 px-4 bg-gray-800 text-white rounded-md focus:outline-none'>
          {wallet.map((acc, i) => (
            <option key={`account-${i}`} className='bg-gray-800' value={acc.address}>
              {`${acc.address.substring(0, 5)}...${acc.address.substring(acc.address.length - 5)}`}
            </option>
          ))}
        </select>
        <div className='flex'>
          <div className='flex flex-col items-center mx-2'>
            <button
              data-testid='send-button'
              onClick={() => setDisplaySend(true)}
              className='bg-blue-500 hover:bg-blue-600 text-white text-l p-2 m-2 rounded-full'>
              <GrTransaction />
            </button>
            <span>Send</span>
          </div>
          <div className='flex flex-col items-center mx-2'>
            <button className='bg-blue-500 hover:bg-blue-600 text-white text-l p-2 m-2 rounded-full'>
              <GrDocumentUpload />
            </button>
            <span>Upload</span>
          </div>
        </div>
      </div>
      {isSendDisplayed ? (
        <SendModal
          web3={web3}
          fromAccount={account}
          onSend={(err) => {
            setDisplaySend(false);
            setErrorMessage(err);
          }}
        />
      ) : null}
      {error ? <ErrorMessage message={error} /> : null}
    </>
  );
}

interface SendProps {
  web3: Web3;
  fromAccount: Wallet.Web3Account;
  onSend: (err?: string) => void;
}

export function SendModal({ web3, fromAccount, onSend }: SendProps) {
  const [toAdress, setAdress] = useState<string>("");
  const [amount, setAmount] = useState<string>("");

  const getErrorMessage = (error: unknown) => (error instanceof Error ? error.message : String(error));

  async function onSubmit(event: FormEvent<HTMLFormElement>) {
    event.preventDefault();

    try {
      const nonce = await web3.eth.getTransactionCount(fromAccount.address, "latest");
      const gasPrice = await web3.eth.getGasPrice();

      const transaction = {
        to: toAdress,
        value: web3.utils.toWei(amount!, "ether"),
        gasPrice: gasPrice,
        nonce: nonce,
      };

      const gas = await web3.eth.estimateGas(transaction);

      const signedTransaction = await fromAccount.signTransaction({
        ...transaction,
        gas,
      });

      await web3.eth.sendSignedTransaction(signedTransaction.rawTransaction);

      onSend();
    } catch (e) {
      onSend(getErrorMessage(e));
    }
  }

  return (
    <>
      <div className='justify-center items-center flex fixed inset-0 z-50 outline-none focus:outline-none'>
        <form
          onSubmit={onSubmit}
          className='p-2 flex flex-col items-around bg-white rounded-md outline-none focus:outline-none'>
          <input
            placeholder='Address'
            type='text'
            className='p-3 m-1 bg-gray-800 text-white rounded-md focus:outline-none'
            onChange={(e) => setAdress(e.target.value)}
            value={toAdress}
          />
          <input
            placeholder='Amount (Eth)'
            type='text'
            className='p-3 m-1 bg-gray-800 text-white rounded-md focus:outline-none'
            onChange={(e) => setAmount(e.target.value)}
            value={amount}
          />
          <button
            className='bg-blue-500 text-white hover:bg-blue-600 font-bold uppercase text-sm p-3 m-1 rounded shadow hover:shadow-lg outline-none focus:outline-none mr-1 mb-1 ease-linear transition-all duration-150'
            type='submit'>
            Send
          </button>
        </form>
      </div>
      <div className='opacity-50 fixed inset-0 z-40 bg-black'></div>
    </>
  );
}
