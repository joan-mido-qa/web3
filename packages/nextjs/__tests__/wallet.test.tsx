import { render, screen, waitFor } from "@testing-library/react";
import Web3Wallet from "../app/wallet/components/Web3Wallet";
import { SendModal } from "../app/wallet/components/Web3Wallet";
import Web3, { WebSocketProvider } from "web3";
import userEvent from "@testing-library/user-event";

function setup(jsx: JSX.Element) {
  return {
    user: userEvent.setup(),
    ...render(jsx),
  };
}

const toNumber = (balance: string) => Number(balance.split(" ")[0]);

let provider: WebSocketProvider;
let web3: Web3;

beforeAll(() => {
  provider = new WebSocketProvider("ws://localhost:8545");
  web3 = new Web3(provider);
  web3.eth.accounts.wallet.add("0x4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d");
  web3.eth.accounts.wallet.add("0x6cbed15c793ce57650b9877cf6fa156fbef513c4e6134f022a85b1ffdd59b2a1");
});

afterAll(async () => {
  await provider.safeDisconnect();
  web3.eth.accounts.wallet.clear();
});

test("renders wallet page unchanged", async () => {
  const { container } = render(<Web3Wallet web3={web3} wallet={web3.eth.accounts.wallet} />);
  expect(container).toMatchSnapshot();
});

test("renders send modal unchanged", async () => {
  const account = web3.eth.accounts.wallet.get(1)!;
  const { container } = render(<SendModal web3={web3} fromAccount={account} onSend={() => {}} />);
  expect(container).toMatchSnapshot();
});

test("send a transaction", async () => {
  const { user } = setup(<Web3Wallet web3={web3} wallet={web3.eth.accounts.wallet} />);

  await waitFor(async () => {
    expect(toNumber(await screen.getByTestId("eth-balance").innerText)).toBeGreaterThan(0);
  });

  const fromInitBalance = toNumber(await screen.getByTestId("eth-balance").innerText);

  await user.selectOptions(screen.getByRole("combobox"), web3.eth.accounts.wallet.get(1)!.address);

  const toInitBalance = toNumber(await screen.getByTestId("eth-balance").innerText);

  await user.click(screen.getByTestId("send-button"));
  await user.type(screen.getByPlaceholderText("Address"), web3.eth.accounts.wallet.get(0)!.address);
  await user.type(screen.getByPlaceholderText("Amount (Eth)"), "1");

  await user.click(
    screen.getByRole("button", {
      name: /send/i,
    })
  );

  await waitFor(async () => {
    expect(toNumber(screen.getByTestId("eth-balance").innerText)).toBeCloseTo(toInitBalance - 1);
  });

  await user.selectOptions(screen.getByRole("combobox"), web3.eth.accounts.wallet.get(0)!.address);

  await waitFor(async () => {
    expect(toNumber(screen.getByTestId("eth-balance").innerText)).toBeCloseTo(fromInitBalance + 1);
  });
});

test("error on send a transaction", async () => {
  const { user } = setup(<Web3Wallet web3={web3} wallet={web3.eth.accounts.wallet} />);

  await user.click(screen.getByTestId("send-button"));
  await user.type(screen.getByPlaceholderText("Address"), "0x04");
  await user.type(screen.getByPlaceholderText("Amount (Eth)"), "1");

  await user.click(
    screen.getByRole("button", {
      name: /send/i,
    })
  );

  await waitFor(async () => {
    expect(screen.getByRole("alert").innerText).toBe("Returned error: The field to must have byte length of 20");
  });
});
