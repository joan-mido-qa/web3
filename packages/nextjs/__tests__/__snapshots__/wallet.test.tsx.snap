// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders send modal unchanged 1`] = `
<div>
  <div
    class="justify-center items-center flex fixed inset-0 z-50 outline-none focus:outline-none"
  >
    <form
      class="p-2 flex flex-col items-around bg-white rounded-md outline-none focus:outline-none"
    >
      <input
        class="p-3 m-1 bg-gray-800 text-white rounded-md focus:outline-none"
        placeholder="Address"
        type="text"
        value=""
      />
      <input
        class="p-3 m-1 bg-gray-800 text-white rounded-md focus:outline-none"
        placeholder="Amount (Eth)"
        type="text"
        value=""
      />
      <button
        class="bg-blue-500 text-white hover:bg-blue-600 font-bold uppercase text-sm p-3 m-1 rounded shadow hover:shadow-lg outline-none focus:outline-none mr-1 mb-1 ease-linear transition-all duration-150"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
  <div
    class="opacity-50 fixed inset-0 z-40 bg-black"
  />
</div>
`;

exports[`renders wallet page unchanged 1`] = `
<div>
  <div
    class="flex flex-col text-white justify-around items-center h-[250px]"
  >
    <span
      class="text-5xl"
      data-testid="eth-balance"
    >
      0
       ETH
    </span>
    <select
      class="m-[5px] py-2 px-4 bg-gray-800 text-white rounded-md focus:outline-none"
    >
      <option
        class="bg-gray-800"
        value="0x90F8bf6A479f320ead074411a4B0e7944Ea8c9C1"
      >
        0x90F...8c9C1
      </option>
      <option
        class="bg-gray-800"
        value="0xFFcf8FDEE72ac11b5c542428B35EEF5769C409f0"
      >
        0xFFc...409f0
      </option>
    </select>
    <div
      class="flex"
    >
      <div
        class="flex flex-col items-center mx-2"
      >
        <button
          class="bg-blue-500 hover:bg-blue-600 text-white text-l p-2 m-2 rounded-full"
          data-testid="send-button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2,7 L20,7 M16,2 L21,7 L16,12 M22,17 L4,17 M8,12 L3,17 L8,22"
              fill="none"
              stroke-width="2"
            />
          </svg>
        </button>
        <span>
          Send
        </span>
      </div>
      <div
        class="flex flex-col items-center mx-2"
      >
        <button
          class="bg-blue-500 hover:bg-blue-600 text-white text-l p-2 m-2 rounded-full"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.99787498,0.999999992 L17.4999998,0.999999992 L20.9999998,4.50000005 L21,23 L3,23 L2.99787498,0.999999992 Z M16,1 L16,6 L21,6 M12,20 L12,11 M8,14 L12,10 L16,14"
              fill="none"
              stroke-width="2"
            />
          </svg>
        </button>
        <span>
          Upload
        </span>
      </div>
    </div>
  </div>
</div>
`;
